replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+
+(* Receive a DELETE request with success target of 2.
+ * For the request to succeed, atleast two replicas
+ *    - must be up AND
+ *    - must have a PUT record for the blob.
+ *)
+RecvDelete(blobId) ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ blobId < blobIdCtr
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ deleteCtr' = deleteCtr + 1
+        /\ [blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r1]
+        /\ [blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r2]
+        /\ [blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r2]
+        /\ [blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r1]
+        /\ storedMessagesPerReplica' = [rr \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((rr = r1) \/ (rr = r2)) THEN
+                storedMessagesPerReplica[rr] \union { [ blobid |-> blobId, blobmessage |-> DELETE ] }
+            ELSE
+                storedMessagesPerReplica[rr]
+            ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, updateTtlCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Receive a UPDATE TTL request with success target of 2.
+ * For the request to succeed, atleast two replicas
+ *    - must be up AND
+ *    - must have a PUT record for the blob AND
+ *    - must not have a DELETE record for the blob.
+ *)
+RecvUpdateTtl(blobId) ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ blobId < blobIdCtr
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r1]
+        /\ [ blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r1]
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] \notin storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] \notin storedMessagesPerReplica[r1]
+        /\ storedMessagesPerReplica' = [rr \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((rr = r1) \/ (rr = r2)) THEN
+                storedMessagesPerReplica[rr] \union { [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] }
+            ELSE
+                storedMessagesPerReplica[rr]
+            ]
+        /\ updateTtlCtr' = updateTtlCtr + 1
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, deleteCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Bring one replica up and one replica down *)
+ReplicaStateChange ==
+    \E m, n \in Replicas:
```

Comment:
This is interesting.  I guess this is ok, but I'm speculating to the utility of having seperate nodes come up and down in the same state.  It might be more clear to break these states up?  Might not make a difference as far as the model checker is concerned.

Reply:
I think you are right. It makes the spec cleaner to break this up. I will make that change.